crates/ruff_linter/src/rules/pylint/rules/useless_else_on_loop.rs

```diff
@@ -91,17 +132,50 @@ fn loop_exits_early(body: &[Stmt]) -> bool {
     })
 }
 
-/// PLW0120
-pub(crate) fn useless_else_on_loop(
-    checker: &mut Checker,
+/// Generate a [`Fix`] to remove the `else` clause from the given statement.
+fn remove_else(
     stmt: &Stmt,
-    body: &[Stmt],
     orelse: &[Stmt],
-) {
-    if !orelse.is_empty() && !loop_exits_early(body) {
-        checker.diagnostics.push(Diagnostic::new(
-            UselessElseOnLoop,
-            identifier::else_(stmt, checker.locator().contents()).unwrap(),
-        ));
+    else_range: TextRange,
+    locator: &Locator,
+    stylist: &Stylist,
+) -> Result<Fix> {
+    let Some(start) = orelse.first() else {
+        return Err(anyhow::anyhow!("Empty `else` clause"));
+    };
+    let Some(end) = orelse.last() else {
+        return Err(anyhow::anyhow!("Empty `else` clause"));
+    };
+
+    let start_indentation = indentation(locator, start);
+    if start_indentation.is_none() {
+        // Inline `else` block (e.g., `else: x = 1`).
+        Ok(Fix::safe_edit(Edit::deletion(
+            else_range.start(),
+            start.start(),
+        )))
+    } else {
+        // Identify the indentation of the loop itself (e.g., the `while` or `for`).
+        let Some(desired_indentation) = indentation(locator, stmt) else {
+            return Err(anyhow::anyhow!("Compound statement cannot be inlined"));
+        };
+
+        // Dedent the content from the end of the `else` to the end of the loop.
+        let indented = adjust_indentation(
+            TextRange::new(
+                locator.full_line_end(else_range.start()),
+                locator.full_line_end(end.end()),
+            ),
+            desired_indentation,
+            locator,
+            stylist,
+        )?;
+
+        // Replace the content from the start of the `else` to the end of the loop.
+        Ok(Fix::safe_edit(Edit::replacement(
+            indented,
+            locator.line_start(else_range.start()),
+            locator.full_line_end(end.end()),
+        )))
```

Comment:
I simplified this a bit, but we now drop comments on the same line as the `else:`. I think this is ok.

Reply:
@diceroll123 - Did you feel strongly about preserving these? I can see why it might be useful, but the comments _do_ get moved, right? Like:

```python
else:  # some comment
  x = 1
  y = 2
```

Was getting moved to:

```python
x = 1
y = 2  # some comment
```

Which struct me as potentially-wrong.

For example, in this case:

```python
def test_retain_comment():
    """Retain the comment within the `else` block"""
    for j in range(10):
        pass
    else:  # [useless-else-on-loop]
        print("fat chance")
        for j in range(10):
            break

```

The comment was moved to after the _break_.

Reply:
(Merging to keep things moving but happy to revisit.)

Reply:
No strong feelings either way! It's much simpler without preserving comments on the else line.